lib/srv/app/server.go

```diff
@@ -591,30 +634,96 @@ func (s *Server) HandleConnection(conn net.Conn) {
 	closerConn.Wait()
 }
 
-func (s *Server) handleConnection(conn net.Conn) error {
+func (s *Server) handleConnection(conn net.Conn) (net.Conn, error) {
+	// Make sure everything here is wrapped in the tracking read connection for monitoring.
+	ctx, cancel := context.WithCancel(s.closeContext)
+	tc, err := srv.NewTrackingReadConn(srv.TrackingReadConnConfig{
+		Conn:    conn,
+		Clock:   s.c.Clock,
+		Context: ctx,
+		Cancel:  cancel,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	// Proxy sends a X.509 client certificate to pass identity information,
 	// extract it and run authorization checks on it.
-	tlsConn, user, app, err := s.getConnectionInfo(s.closeContext, conn)
+	tlsConn, user, app, err := s.getConnectionInfo(ctx, tc)
 	if err != nil {
-		return trace.Wrap(err)
+		return tlsConn, trace.Wrap(err)
+	}
+	authCtx, _, err := s.authorizeContext(context.WithValue(ctx, auth.ContextUser, user))
+	if err != nil {
+		if !app.IsTCP() {
+			s.setConnAuth(tlsConn, err)
+		} else {
+			return tlsConn, trace.Wrap(err)
+		}
+	} else {
+		err = s.monitorConn(ctx, tc, authCtx)
+		if err != nil {
+			return tlsConn, trace.Wrap(err)
+		}
 	}
 
 	// Application access supports plain TCP connections which are handled
 	// differently than HTTP requests from web apps.
 	if app.IsTCP() {
-		return s.handleTCPApp(s.closeContext, tlsConn, user, app)
+		identity := authCtx.Identity.GetIdentity()
+		return tlsConn, s.handleTCPApp(ctx, tlsConn, &identity, app)
 	}
 
-	return s.handleHTTPApp(s.closeContext, tlsConn)
+	return tlsConn, s.handleHTTPApp(ctx, tlsConn)
 }
 
-// handleTCPApp handles connection for a TCP application.
-func (s *Server) handleTCPApp(ctx context.Context, conn net.Conn, user auth.IdentityGetter, app types.Application) error {
-	id, _, err := s.authorizeContext(context.WithValue(ctx, auth.ContextUser, user))
+// monitorConn takes a TrackingReadConn and starts a connection monitor. The tracking connection will be
+// auto-terminated if disconnect_expired_cert or idle timeout is configured, and unmodified client connection
+// otherwise.
```

Comment:
Nit: This godoc reads weird. "The connection will be auto-terminated ..., and unmodified client connection otherwise". Feels like the second part of the sentence doesn't correspond to the first one :)

Reply:
Bad copy paste. :-X